### PR TITLE
✨ Feature: `wkb` Decoding via `numpy`

### PR DESCRIPTION
**Pull request by measty**
This PR adds the option for geometries to be returned as [wkb](https://www.ibm.com/docs/en/db2-warehouse?topic=formats-well-known-binary-wkb-format), and provides an efficient decoder that returns the coordinates list directly from the wkb. This is significantly more efficient (2-3x faster) if you only need the contours, as everything is not converted through shapely first as an intermediary.

Also fixes a few docstrings and other inconsistencies such as liskov substitution for some query functions of AnnotationStore and SQLiteStore.

### CODE REVIEW

File: tiatoolbox/utils/visualization.py [hunk 1]
```diff
@@ -12,9 +12,19 @@
 from PIL import Image, ImageFilter, ImageOps
 from shapely.geometry import Polygon
 
-from tiatoolbox import logger
+from tiatoolbox import DuplicateFilter, logger
 from tiatoolbox.annotation.storage import Annotation, AnnotationStore
 
+GEOMTYPES = {
+    1: "Point",
```
Comment (shaneahmed) [inline tiatoolbox/utils/visualization.py:19]: shouldn't it start from 0?
```suggestion
    0: "Point",
```

File: tiatoolbox/utils/visualization.py [hunk 2]
```diff
@@ -534,6 +544,14 @@ class AnnotationRenderer:
             will override the primary colormap.
         blur_radius (int):
             radius of gaussian blur to apply to rendered annotations.
+        score_prop_edge (str):
+            A key that is present in the properties of annotations
+            to be rendered that will be used to color rendered edges.
+        function_mapper (Callable):
+            a callable which will be given the properties of an annotation
```
Comment (shaneahmed) [inline tiatoolbox/utils/visualization.py:551]:
```suggestion
            A callable which will be given the properties of an annotation
```

File: tiatoolbox/utils/visualization.py [hunk 3]
```diff
@@ -573,12 +592,84 @@ def __init__(
...
             self.blur = ImageFilter.GaussianBlur(blur_radius)
             self.edge_thickness = 0
         else:
             self.blur = None
 
+    @staticmethod
+    def decode_geom(geom, geom_type):
+        """Decode a geometry from the raw wkb bytes representation
+        to an array of coordinates.
+        Args:
```
Comment (shaneahmed) [inline tiatoolbox/utils/visualization.py:606]:
```suggestion
        
        Args:
```

File: tiatoolbox/utils/visualization.py [hunk 4]
```diff
@@ -573,12 +592,84 @@ def __init__(
...
+        self.function_mapper = function_mapper
         if blur_radius > 0:
             self.blur = ImageFilter.GaussianBlur(blur_radius)
             self.edge_thickness = 0
         else:
             self.blur = None
 
+    @staticmethod
+    def decode_geom(geom, geom_type):
+        """Decode a geometry from the raw wkb bytes representation
```
Comment (shaneahmed) [inline tiatoolbox/utils/visualization.py:604]: First line should be a brief sentence describing the function.

File: tiatoolbox/utils/visualization.py [hunk 5]
```diff
@@ -573,12 +592,84 @@ def __init__(
...
+        to an array of coordinates.
+        Args:
+            geom (bytes):
+                The raw wkb bytes representation of a geometry.
+            type (str):
+                The type of geometry.
+        Returns:
```
Comment (shaneahmed) [inline tiatoolbox/utils/visualization.py:611]:
```suggestion
        
        Returns:
```

File: tiatoolbox/utils/visualization.py [hunk 6]
```diff
@@ -573,12 +592,84 @@ def __init__(
...
             self.blur = None
 
+    @staticmethod
+    def decode_geom(geom, geom_type):
+        """Decode a geometry from the raw wkb bytes representation
+        to an array of coordinates.
+        Args:
+            geom (bytes):
+                The raw wkb bytes representation of a geometry.
+            type (str):
```
Comment (shaneahmed) [inline tiatoolbox/utils/visualization.py:609]:
```suggestion
            geom_type (GEOMTYPES):
```

File: tiatoolbox/utils/visualization.py [hunk 7]
```diff
@@ -819,7 +940,7 @@ def render_annotations(
             for i, (key, box) in enumerate(bounding_boxes.items()):
                 area = (box[0] - box[2]) * (box[1] - box[3])
                 if area > min_area or i % decimate == 0:
-                    ann = store[key]
+                    ann = store.__getitem__(key, as_wkb=True)
```
Comment (shaneahmed) [inline tiatoolbox/utils/visualization.py:943]:
```suggestion
                    ann = store[key, True]
```

File: tiatoolbox/annotation/storage.py [hunk 8]
```diff
@@ -144,6 +144,16 @@ def to_geojson(self) -> str:
...
         return json.dumps(self.to_feature())
 
+    def to_wkb_geometry(self):
+        """
+        Return the annotation with a WKB geometry.
+
+        Returns:
+            Annotation:
+                The annotation with a WKB geometry.
+        """
```
Comment (shaneahmed) [inline tiatoolbox/annotation/storage.py:154]:
```suggestion
        
        """
```

File: tiatoolbox/annotation/storage.py [hunk 9]
```diff
@@ -144,6 +144,16 @@ def to_geojson(self) -> str:
         """
         return json.dumps(self.to_feature())
 
+    def to_wkb_geometry(self):
+        """
+        Return the annotation with a WKB geometry.
```
Comment (shaneahmed) [inline tiatoolbox/annotation/storage.py:149]:
```suggestion
        """Returns the annotation with a WKB geometry.
```

Comment (shaneahmed) [reply tiatoolbox/annotation/storage.py:149]: Please add some description about WKB with a link if possible.

File: tests/test_annotation_tilerendering.py [hunk 10]
```diff
@@ -328,10 +328,10 @@ def test_categorical_mapper(fill_store, tmp_path):
             assert 0 <= val <= 1
 
 
-def test_colour_prop_warning(fill_store, tmp_path, caplog):
+def test_colour_prop_warnings(fill_store, tmp_path, caplog):
     """
     Test warning when rendering annotations in which the provided
```
Comment (shaneahmed) [inline tests/test_annotation_tilerendering.py:333]:
```suggestion
    """Test warning when rendering annotations in which the provided
```

Comment (shaneahmed) [reply tests/test_annotation_tilerendering.py:333]: First line of docstring should be a sentence briefly describing the function.

File: tests/test_annotation_tilerendering.py [hunk 11]
```diff
@@ -328,10 +328,10 @@ def test_categorical_mapper(fill_store, tmp_path):
...
-    score_prop does not exist.
+    score_prop does not exist, or its type doesnt match the mapper.
```
Comment (shaneahmed) [inline tests/test_annotation_tilerendering.py:334]:
```suggestion
    score_prop does not exist, or its type does not match the mapper.
```

File: tests/test_annotation_tilerendering.py [hunk 12]
```diff
@@ -342,6 +342,11 @@ def test_colour_prop_warning(fill_store, tmp_path, caplog):
     tg.get_tile(1, 0, 0)
     assert "not found in properties" in caplog.text
 
+    renderer = AnnotationRenderer(score_prop="type", mapper="jet")
+    tg = AnnotationTileGenerator(wsi.info, store, renderer, tile_size=256)
+    tg.get_tile(1, 0, 0)
+    assert "property value type incompatable" in caplog.text
```
Comment (shaneahmed) [inline tests/test_annotation_tilerendering.py:348]:
```suggestion
    assert "property value type incompatible" in caplog.text
```

File: tests/test_annotation_tilerendering.py [hunk 13]
```diff
@@ -420,3 +425,28 @@ def test_multipolygon_render(cell_grid, tmp_path):
...
+def test_function_mapper(fill_store, tmp_path):
+    """Test function mapper."""
+    array = np.ones((1024, 1024))
+    wsi = wsireader.VirtualWSIReader(array, mpp=(1, 1))
+    _, store = fill_store(SQLiteStore, tmp_path / "test.db")
+
+    def color_fn(props):
+        # simple test function that returns red for cells, otherwise green.
+        if props["type"] == "cell":
+            return (1, 0, 0)
```
Comment (shaneahmed) [inline tests/test_annotation_tilerendering.py:439]:
```suggestion
            return 1, 0, 0
```

File: tests/test_annotation_tilerendering.py [hunk 14]
```diff
@@ -420,3 +425,28 @@ def test_multipolygon_render(cell_grid, tmp_path):
...
+        return (0, 1, 0)
```
Comment (shaneahmed) [inline tests/test_annotation_tilerendering.py:440]:
```suggestion
        return 0, 1, 0
```

File: tests/test_visualization.py [hunk 15]
```diff
@@ -189,3 +198,66 @@ def test_plot_graph():
         edges,
     )
     plot_graph(canvas, nodes, edges, node_colors=node_colors, edge_colors=edge_colors)
+
+
+def test_decode_geom():
+    renderer = AnnotationRenderer()
```
Comment (shaneahmed) [inline tests/test_visualization.py:204]: Add docstring.

File: tests/test_visualization.py [hunk 16]
```diff
@@ -189,3 +198,66 @@ def test_plot_graph():
...
+    polygon_contours = renderer.decode_geom(polygon_wkb, 3).reshape(-1, 2)
+
+    # Check that the decoded contours are as expected
+    assert np.all(point_contours == np.array([[0, 0]]))
+    assert np.all(line_contours == np.array([[0, 0], [1, 1], [2, 0]]))
+    assert np.all(
+        polygon_contours == np.array([[0, 0], [0, 1], [1, 1], [1, 0], [0, 0]])
+    )
+
+    # do the same for multi-point, multi-line and multi-polygon
```
Comment (shaneahmed) [inline tests/test_visualization.py:228]:
```suggestion
    # do the same for multipoint, multi-line and multi-polygon
```

File: tiatoolbox/annotation/storage.py [hunk 17]
```diff
@@ -1115,6 +1137,9 @@ def nquery(
                 of two strings. The first string is the mode for the
                 query geometry and the second string is the mode for
                 the nearest annotation geometry.
+            as_wkb (bool):
+                If True, return geometries as raw wkb bytes instead of
+                shapely geometries. Defaults to False.
 
         Returns:
             Dict[str, Dict[str, Annotation]]:
```
Comment (shaneahmed) [inline tiatoolbox/annotation/storage.py:1145]: line 1184 --> Polygon

File: tiatoolbox/utils/visualization.py [hunk 18]
```diff
@@ -594,13 +685,16 @@ def to_tile_coords(coords: List, top_left: Tuple[float, float], scale: float):
...
 
     def get_color(self, annotation: Annotation, edge=False):
         """Get the color for an annotation.
         Args:
             annotation (Annotation):
                 Annotation to get color for.
+            edge (bool):
+                Whether to get the color for the edge of the annotation,
+                or the interior.
         Returns:
```
Comment (shaneahmed) [inline tiatoolbox/utils/visualization.py:698]:
```suggestion
        
        Returns:
```

File: tiatoolbox/utils/visualization.py [hunk 19]
```diff
@@ -769,17 +880,26 @@ def render_annotations(
...
+                The bounding box of the tile to render.
             scale (float):
                 The scale at which we are rendering the tile.
+            res (int):
+                The resolution of the tile. Defaults to 1. Can be set to 2 for
+                higher resolution rendering.
+            border (int):
+                The border to add around the tile. Defaults to 0. Used for blurred
+                rendering to avoid edge effects.
         Returns:
```
Comment (shaneahmed) [inline tiatoolbox/utils/visualization.py:895]:
```suggestion
        
        Returns:
```

File: tiatoolbox/utils/visualization.py [hunk 20]
```diff
@@ -769,17 +880,26 @@ def render_annotations(
...
             np.ndarray:
```
Comment (shaneahmed) [inline tiatoolbox/utils/visualization.py:896]:
```suggestion
            :class:`np.ndarray`:
```

**Review by shaneahmed — COMMENTED** [review]
Thanks @measty. I have added my comments.

File: tiatoolbox/utils/visualization.py (hunk 1 above)
Comment (measty) [reply tiatoolbox/utils/visualization.py:19]: No, in the wkb specification it starts from 1. See:
https://www.ibm.com/docs/en/db2-warehouse?topic=formats-well-known-binary-wkb-format

File: tiatoolbox/utils/visualization.py [hunk 21]
```diff
@@ -573,12 +592,84 @@ def __init__(
...
         self.blur_radius = blur_radius
+        self.function_mapper = function_mapper
         if blur_radius > 0:
             self.blur = ImageFilter.GaussianBlur(blur_radius)
             self.edge_thickness = 0
         else:
             self.blur = None
 
+    @staticmethod
+    def decode_geom(geom, geom_type):
```
Comment (John-P) [inline tiatoolbox/utils/visualization.py:603]:
```suggestion
    def decode_geom(geom: bytes, geom_type: int) -> np.array:
```

Comment (John-P) [inline tiatoolbox/utils/visualization.py:603]:
```suggestion
    def decode_wkb(wkb, geom_type):
```

Maybe makes it more clear.

**Review by John-P — COMMENTED** [review]
Is it faster than shapely? I am just curious as the shapely methods are in C/C++. Great if it is! We could even use numba to make it faster.

One problem I can see at the moment is that this is breaking the typing with Annotation. We should either update the Annotation typing to indicate that geometry can be a Geometry OR a NumPy array, or else we can have a new NpAnnotation (or some other name) object where the geometry is always raw NumPy array.

I worry that having too many switches and multiple different possible types for the return values may get confusing and lead to user errors/bugs. However, I guess this flag wouldn't be used unless you are an advanced user and have read the docs.

**Comment by shaneahmed** [comment]
In https://github.com/TissueImageAnalytics/tiatoolbox/blob/fe1c80905ed3c593b5e93e75707057dc8685668f/tiatoolbox/annotation/storage.py#L1675, there are no attributes connection or path. Please can you check.

File: tests/test_visualization.py [hunk 22]
```diff
@@ -189,3 +198,66 @@ def test_plot_graph():
...
+    polygon_contours = renderer.decode_geom(polygon_wkb, 3).reshape(-1, 2)
+
+    # Check that the decoded contours are as expected
+    assert np.all(point_contours == np.array([[0, 0]]))
+    assert np.all(line_contours == np.array([[0, 0], [1, 1], [2, 0]]))
+    assert np.all(
+        polygon_contours == np.array([[0, 0], [0, 1], [1, 1], [1, 0], [0, 0]])
+    )
+
+    # do the same for multipoint, multiline and multipolygon
```
Comment (shaneahmed) [inline tests/test_visualization.py:228]:
```suggestion
    # do the same for multi-point, multi-line and multi-polygon
```

**Comment by measty** [comment]
I've address all the minor docstring/naming/formatting comments, the outstanding comments/issue still to be resolved now are connected to how best to resolve the typing issue with Annotation.

File: tiatoolbox/utils/visualization.py (hunk 7 above)
Comment (shaneahmed) [reply tiatoolbox/utils/visualization.py]: Will be fix in #639

**Comment by shaneahmed** [comment]
> I've address all the minor docstring/naming/formatting comments, the outstanding comments/issue still to be resolved now are connected to how best to resolve the typing issue with Annotation.

Just resolving conflicts before merge.

**Review by shaneahmed — APPROVED** [review]

**Pull request merged**